File: tracker-updater.js
#!/usr/bin/env node

const async = require("async");
const bz = require("bz");
const colors = require("colors/safe");
const RT = require("bestpractical-rt");

const options = require("node-getopt-long").options(
  [
    ["rt_url|rt-url=s", "BWS RT URL"],
    ["rt_username|rt-username=s", "BWS RT username"],
    ["rt_password|rt-password=s", "BWS RT password"],
    ["dev_tracker_url|dev-url=s", "BWS tracker URL"],
    ["dev_tracker_username|dev-username=s", "BWS tracker username"],
    ["dev_tracker_password|dev-password=s", "BWS tracker password"],
    ["community_url|community-url=s", "Community tracker URL"],
    ["community_username|community-username=s", "Community tracker username"],
    ["community_password|community-password=s", "Community tracker password"],
    ["interactive|i", "Interactive mode"]
  ],
  {
    name: "tracker-updater",
    commandVersion: 1.0,
    defaults: {
      dev_tracker_url: "http://tracker.devs.bywatersolutions.com/rest/",
      community_url: "https://bugs.koha-community.org/bugzilla3/rest/"
    }
  }
);

const rt = new RT(options.rt_username, options.rt_password, options.rt_url);

const bws_tracker = bz.createClient({
  url: options.dev_tracker_url,
  username: options.dev_tracker_username,
  password: options.dev_tracker_password,
  timeout: 30000
});

const community_tracker = bz.createClient({
  url: options.community_url,
  username: options.community_username,
  password: options.community_password,
  timeout: 30000
});

process();

async function process() {
  await create_tracks();
  await create_community_bugs();
  await update_community_bugs();
}

function create_tracks() {
  return new Promise(function(resolve, reject) {
    const rt_query =
      "Queue = 'Development' AND CF.{Workflow} LIKE 'In Development' AND 'CF.{Work to be done}' IS NOT NULL AND 'CF.{Dev Tracker}' IS NULL";

    rt.search(rt_query, function(results) {
      console.log(
        colors.black.bgWhite("Checking RT for tickets that need Dev Tracks...")
      );

      Object.keys(results).forEach(async function(rt_ticket_id) {
        const val = results[rt_ticket_id];
        console.log(rt_ticket_id + " => " + val);

        // Wait a random number of seconds so we don't kill RT
        const min = 1;
        const max = 5;
        const rand = Math.floor(Math.random() * (max - min + 1) + min);
        await sleep(rand);

        rt.ticketProperties(rt_ticket_id, function(rt_ticket) {
          // These should never trigger because of the search limits we have already
          if (!rt_ticket["CF.{Work to be done}"]) return;
          if (rt_ticket["CF.{Dev Tracker}"]) return;

          const product = rt_ticket["CF.{Development Type}"] == "LibKi" // Mis-capitalized in RT
            ? "Libki"
            : "Koha";
          const component = rt_ticket["CF.{Development Type}"] == "Koha"
            ? "General"
            : rt_ticket["CF.{Development Type}"] == "LibKi" // Mis-capitalized in RT
              ? "General"
              : rt_ticket["CF.{Development Type}"];
          const version = rt_ticket["CF.{Development Type}"] == "LibKi" // Mis-capitalized in RT
            ? "Libki 2"
            : "unspecified";

          const track_data = {
            product: product,
            component: component,
            version: version,
            assigned_to: "jesse@bywatersolutions.com",
            summary: rt_ticket["Subject"],
            description: rt_ticket["CF.{Work to be done}"],
            op_sys: "All",
            rep_platform: "All",
            cf_rt_ticket: rt_ticket_id,
            bug_status: "In Development"
          };

          bws_tracker.createBug(track_data, function(error, track_id) {
            if (error) {
              console.log(
                colors.red("ERROR CREATING BUG REPORT ON DEV TRACKER!")
              );
              console.log(error);
              process.exit(1);
            }
            console.log("TRACK ID: " + track_id);

            // Update RT ticket with track id
            const properties = { "CF-Dev Tracker": track_id };
            rt.updateTicketProperties(rt_ticket_id, properties, function() {
              console.log("RT TICKET UPDATED");
            });
          });
        });
      });

      resolve();
    });
  });
}

function create_community_bugs() {
  return new Promise(function(resolve, reject) {
    let searchParams = {
      status: "In Development"
    };

    bws_tracker.searchBugs(searchParams, function(error, bugs) {
      console.log(
        colors.black.bgWhite(
          "Checking for Dev Tracks that need Community Bugs..."
        )
      );

      if (error) {
        console.log(colors.red("ERROR SEARCHING BWS DEV TRACKER!"));
        console.log(error);
        process.exit(1);
      }

      for (let i = 0; i < bugs.length; i++) {
        const bws_bug = bugs[i];

        // If there is already a community bug, we don't want to create a duplicate
        if (bws_bug.cf_community_bug) continue;

        // We don't want to mess with other projects
        if (bws_bug.product != "Koha") continue;

        // Plugins and XSLT mods don't get submitted to the community
        if (bws_bug.component != "General") continue;

        bws_tracker.bugComments(bws_bug.id, function(error, comments) {
          if (error) {
            console.log(
              colors.red("ERROR GETTING COMMENTS FROM COMMUNITY TRACKER!")
            );
            console.log(error);
            process.exit(1);
          }

          const comment = comments[0].text;

          const community_bug_data = {
            product: "Koha",
            component: "Architecture, internals, and plumbing",
            version: "master",
            assigned_to: bws_bug.assigned_to,
            summary: bws_bug.summary,
            description: comment
          };

          community_tracker.createBug(community_bug_data, function(
            error,
            community_bug_id
          ) {
            if (error) {
              console.log(
                colors.red("ERROR CREATING BUG REPORT ON COMMUNITY TRACKER!")
              );
              console.log(error);
              process.exit(1);
            }

            bws_tracker.updateBug(
              bws_bug.id,
              {
                cf_community_bug: community_bug_id
              },
              function(error, updated_bug) {
                if (error) {
                  console.log(
                    colors.red(
                      "ERROR ADDING NEW COMMUNITY BUG ID TO BWS DEV TRACKER!"
                    )
                  );
                  console.log(error);
                  process.exit(1);
                }

                console.log(
                  "Created community bug " +
                    colors.cyan(community_bug_id) +
                    " for tracker bug " +
                    colors.green(bws_bug.id)
                );
              }
            );
          });
        });
      }

      resolve();
    });
  });
}

function update_community_bugs() {
  return new Promise(function(resolve, reject) {
    searchParams = { status: "Submitted to Community" };
    bws_tracker.searchBugs(searchParams, function(error, bugs) {
      console.log(
        colors.black.bgWhite("Updating Dev Tracks from Community Bugs...")
      );

      if (error) {
        console.log(colors.red("ERROR SEARCHING BWS DEV TRACKER!"));
        console.log(error);
        process.exit(1);
      }

      for (let i = 0; i < bugs.length; i++) {
        let bws_bug = bugs[i];
        let bws_status = bws_bug.cf_community_status;

        // Set the community bug id to 0 to skip community processing
        if (bws_bug.cf_community_bug == "0") continue;

        community_tracker.getBug(bws_bug.cf_community_bug, function(
          error,
          community_bug
        ) {
          if (error) {
            console.log("BWS ID: " + colors.red(bws_bug.id));
            console.log(colors.red("ERROR GETTING COMMUNITY BUG!"));
            console.log(error);
            return;
          }

          let community_status = community_bug.status;
          let community_summary = community_bug.summary;

          console.log("BWS ID: " + colors.green(bws_bug.id));
          console.log("BWS STATUS: " + colors.green(bws_status));
          console.log("STATUS: " + colors.cyan(community_status));

          if (bws_status != community_status) {
            bws_tracker.updateBug(
              bws_bug.id,
              {
                cf_community_status: community_status,
                summary: community_summary
              },
              function(error, ok) {
                console.log(
                  `Updating BWS Tracker Bug ${colors.green(
                    bws_bug.id
                  )} with status ${colors.green(
                    bws_status
                  )} to community status ${colors.cyan(
                    community_status
                  )} from community bug ${colors.cyan(community_bug.id)}`
                );
                if (error) {
                  console.log(colors.red("ERROR UPDATING BWS TRACKER!"));
                  console.log(error);
                }
              }
            );
          }
        });
      }
    });
    resolve();
  });
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms * 100));
}

File: tracker-updater.pl
#!/usr/bin/env perl

use Modern::Perl;

use BZ::Client::REST;
use Carp::Always;
use Data::Dumper;
use Getopt::Long::Descriptive;
use JSON qw(to_json);
use LWP::UserAgent;
use RT::Client::REST;
use Term::ANSIColor;
use Try::Tiny;

my ( $opt, $usage ) = describe_options(
    'tracker-updater.pl',
    [ "rt-url=s",      "BWS RT URL",      { required => 1, default => $ENV{RT_URL} } ],
    [ "rt-username=s", "BWS RT username", { required => 1, default => $ENV{RT_USER} } ],
    [ "rt-password=s", "BWS RT password", { required => 1, default => $ENV{RT_PW} } ],
    [],
    [ "dev-url=s",      "BWS tracker URL",      { required => 1, default => $ENV{BWS_URL} } ],
    [ "dev-username=s", "BWS tracker username", { required => 1, default => $ENV{BWS_USER} } ],
    [ "dev-password=s", "BWS tracker password", { required => 1, default => $ENV{BWS_PW} } ],
    [],
    [ "community-url=s",      "Community tracker URL",      { required => 1, default => $ENV{KOHA_URL} } ],
    [ "community-username=s", "Community tracker username", { required => 1, default => $ENV{KOHA_USER} } ],
    [ "community-password=s", "Community tracker password", { required => 1, default => $ENV{KOHA_PW} } ],
    [],
    [ "action|a=s",    "Optional action to perform ( create-track )" ],
    [ "bug|b=s",       "Bug to perform action on" ],
    [ "dev-track|d=s", "Track to perform action on" ],
    [ "ticket|t=s",    "RT Ticket to perform action on" ],
    [],
    [ 'slack|s=s', "Slack webhook URL", { required => 1, default => $ENV{SLACK_URL} } ],
    [],
    [ "force|f", "Get pushy" ],
    [ 'verbose|v+', "Print extra stuff", { required => 1, default => 0 } ],
    [ 'help|h', "Print usage message and exit", { shortcircuit => 1 } ],
);

print( $usage->text ), exit if $opt->help;

my $rt_url  = $opt->rt_url;
my $rt_user = $opt->rt_username;
my $rt_pass = $opt->rt_password;

my $bz_tracker_url  = $opt->dev_url;
my $bz_tracker_user = $opt->dev_username;
my $bz_tracker_pass = $opt->dev_password;

my $bz_koha_url  = $opt->community_url;
my $bz_koha_user = $opt->community_username;
my $bz_koha_pass = $opt->community_password;

my $tracker_client = BZ::Client::REST->new(
    {
        user     => $bz_tracker_user,
        password => $bz_tracker_pass,
        url      => $bz_tracker_url,
    }
);

my $koha_client = BZ::Client::REST->new(
    {
        user     => $bz_koha_user,
        password => $bz_koha_pass,
        url      => $bz_koha_url,
    }
);

my $rt = RT::Client::REST->new(
    server  => $rt_url,
    timeout => 30,
);
try {
    $rt->login( username => $rt_user, password => $rt_pass );
}
catch {
    die "problem logging in: ", shift->message;
};

my $ua = LWP::UserAgent->new;
$ua->post(
    $opt->slack,
    Content_Type => 'application/json',
    Content => to_json( { text => "Running dev tracker updater!" } ),
) if $opt->slack;

if ( $opt->action ) {
    if ( $opt->action eq 'create-track' ) {
        my $bug_id = $opt->bug;

        my $results =
          $tracker_client->search_bugs( { cf_community_bug => $bug_id } );
        my $track_id;
        foreach my $r (@$results) {
            if ( $r->{status} ne 'RESOLVED' ) {
                $track_id = $r->{id};
                last;
            }
        }

        if ( $track_id && !$opt->force ) {
            say 'Track '
              . colored( $track_id, 'red' )
              . ' already exists for bug '
              . colored( $bug_id, 'cyan' );

        }
        else {
            my $bug = $koha_client->get_bug($bug_id);

            my $track_data = {
                product             => 'Koha',
                component           => 'General',
                version             => 'unspecified',
                assigned_to         => $bz_tracker_user,
                summary             => $bug->{summary},
                op_sys              => 'All',
                rep_platform        => 'All',
                cf_community_bug    => $bug_id,
                cf_community_status => $bug->{status},
                bug_status          => 'Submitted to Community',
            };

            my $track_id = $tracker_client->create_bug($track_data);

            say 'Created track '
              . colored( $track_id, 'green' )
              . ' for bug '
              . colored( $bug_id, 'cyan' );

            $ua->post(
                $opt->slack,
                Content_Type => 'application/json',
                Content => to_json( { text => "Created <$bz_tracker_url/show_bug.cgi?id=$track_id|Track $track_id> for <$bz_koha_url/show_bug.cgi?id=$bug_id|Bug $bug_id>" } ),
            ) if $opt->slack;
        }
    }
    exit 0;
}

# Create tracks
say colored( 'Creating Tracks', 'green' );
my $rt_query =
"Queue = 'Development' AND CF.{Workflow} LIKE 'In Development' AND 'CF.{Work to be done}' IS NOT NULL AND 'CF.{Dev Tracker}' IS NULL";
my @ids = $rt->search(
    type    => 'ticket',
    query   => $rt_query,
    orderby => '-id',
);

my @tickets_needing_tracks;

foreach my $id (@ids) {
    sleep(1);    # pause for 1 second between requests so we don't kill RT
    my $ticket = $rt->show( type => 'ticket', id => $id );
    $ticket->{id} = $id;
    push( @tickets_needing_tracks, $ticket );
    say "Found ticket: " . colored( $id, 'cyan' );
}

foreach my $t (@tickets_needing_tracks) {
    next if !$t->{'CF.{Work to be done}'};
    next if $t->{'CF.{Dev Tracker}'};

    my $product =
      $t->{'CF.{Development Type}'} eq 'LibKi'    # Mis-capitalized in RT
      ? 'Libki'
      : 'Koha';
    my $component =
        $t->{'CF.{Development Type}'} eq 'Koha'  ? 'General'
      : $t->{'CF.{Development Type}'} eq 'LibKi' ? 'General'
      :   $t->{'CF.{Development Type}'};
    my $version =
      $t->{'CF.{Development Type}'} eq 'LibKi'
      ? 'Libki 2'
      : 'unspecified';

    my $track_data = {
        product      => $product,
        component    => $component,
        version      => $version,
        assigned_to  => 'jesse@bywatersolutions.com',
        summary      => $t->{'Subject'},
        description  => $t->{'CF.{Work to be done}'},
        op_sys       => 'All',
        rep_platform => 'All',
        cf_rt_ticket => $t->{id},
        bug_status   => 'Ready for Development'
    };

    my $track_id = $tracker_client->create_bug($track_data);
    say 'Created track: ' . colored( $track_id, 'green' );

    $ua->post(
        $opt->slack,
        Content_Type => 'application/json',
        Content => to_json( { text => "Created <$bz_tracker_url/show_bug.cgi?id=$track_id|Track $track_id> for <$rt_url/Ticket/Display.html?id=$t->{id}|Ticket $t->{id}>" } ),
    ) if $opt->slack;

    $rt->edit(
        type => 'ticket',
        id   => $t->{id},
        set  => { "CF.{Dev Tracker}" => $track_id }
    );
}

# Update tracks from community bugs
say colored( 'Updating Tracks from Community Bugs', 'green' );
my $results = $tracker_client->search_bugs(
    {
        status => [
            'Submitted to Community',
            'In Development',
            'Ready for Development',
        ]
    }
);
foreach my $track ( @$results ) {
    next if $track->{component} eq 'Plugin'; # Plugins exist outside community process
    next if $track->{product} ne 'Koha'; # Community process is only used for Koha
    $track->{cf_community_bug}    ||= q{};
    $track->{cf_koha_version}     ||= q{};
    $track->{cf_community_status} ||= q{};

    ( $track->{cf_community_bug} ) = split( / /, $track->{cf_community_bug} ); # Only use leftmost bug if multiple bugs are associated with track

    next unless $track->{cf_community_bug};

    say "Found track: " . colored( $track->{id}, 'cyan' ) if $opt->verbose;

    my $bug = $koha_client->get_bug( $track->{cf_community_bug} );
    say "Bug data: " . Data::Dumper::Dumper( $bug ) if $opt->verbose > 2;
    $bug->{status} ||= q{};

    if ( $track->{cf_community_status} ne $bug->{status} ) {
        $tracker_client->update_bug( $track->{id}, { cf_community_status => $bug->{status} } );

        my $json_data = {
            "attachments" => [
                {
                    title => "Updated <$bz_tracker_url/show_bug.cgi?id=$track->{id}|Track $track->{id}>",
                    #pretext => "Pretext _supports_ mrkdwn",
                    text => "<$bz_koha_url/show_bug.cgi?id=$bug->{id}|Boog $bug->{id}: $bug->{summary}>",
                    fields => [
                        {
                            title => "From",
                            value => $track->{cf_community_status},
                            short => JSON::true,
                        },
                        {
                            title => "To",
                            value => $bug->{status},
                            short => JSON::true,
                        }
                    ],
                    mrkdwn_in => [ "text", "pretext", "fields" ],
                }
            ]
        };
        my $json_text = to_json( $json_data );

        say 'Updated track ' . colored( $track->{id}, 'cyan' ) . ': ' . colored( $track->{cf_community_status}, 'red' ) . ' => ' . colored( $bug->{status}, 'green' );
        $ua->post(
            $opt->slack,
            Content_Type => 'application/json',
            Content      => $json_text,
        ) if $opt->slack;
    }

    my @tickets = split( / /, $track->{cf_rt_ticket} );
    foreach my $ticket (@tickets) {
	next unless $ticket;

	say "Updating ticket " . colored( $ticket, 'magenta' ) . " for track " . colored( $track->{id}, 'cyan' ) if $opt->verbose > 1;

        try {
            $rt->edit(
                type => 'ticket',
                id   => $ticket,
                set  => {
                    "CF.{Community Status}" => $track->{cf_community_status},
                }
            );
        };

        try {
            $rt->edit(
                type => 'ticket',
                id   => $ticket,
                set  => {
                    "CF.{Koha Version}" => $track->{cf_koha_version},
                }
            );
        };

        try {
            $rt->edit(
                type => 'ticket',
                id   => $ticket,
                set  => {
                    "CF.{Dev Tracker}" => $track->{id},
                }
            );
        }
    };

    sleep(5);    # pause for 10 second between requests so we don't kill RT and/or Bugzilla
}

$ua->post(
    $opt->slack,
    Content_Type => 'application/json',
    Content => to_json( { text => "Dev tracker updater has finished running!" } ),
) if $opt->slack;
say colored( 'Finished!', 'green' );
